File: src/core/scala/mrtjp/core/fx/SequenceAction.scala
/*
 * Copyright (c) 2015.
 * Created by MrTJP.
 * All rights reserved.
 */
package mrtjp.core.fx

import mrtjp.core.fx.particles.CoreParticle

import scala.collection.mutable.{Seq => MSeq}

class SequenceAction extends ParticleAction
{
    var actions = MSeq[ParticleAction]()

    override def tickLife()
    {
        super.tickLife()
        actions.find(!_.isFinished) match
        {
            case Some(action) => action.tickLife()
            case None =>
        }
    }

    override def runOn(p:CoreParticle, frame:Float)
    {
        super.runOn(p, frame)

        actions.find(!_.isFinished) match
        {
            case Some(action) => action.runOn(p, frame)
            case None =>
                isFinished = true
                return
        }

        if (actions.forall(_.isFinished))
            isFinished = true
    }

    override def operate(p:CoreParticle, time:Double){}

    override def compile(p:CoreParticle)
    {
        super.compile(p)
        actions.foreach(_.compile(p))
    }

    override def reset()
    {
        super.reset()
        actions.foreach(_.reset())
    }

    override def copy = ParticleAction.sequence(actions.map(_.copy).toList:_*)
}

File: core/src/main/java/mrtjp/projectred/core/client/particle/SequenceAction.java
package mrtjp.projectred.core.client.particle;

public class SequenceAction extends ParticleAction {

    private final ParticleAction[] actions;
    private int index = 0;

    public SequenceAction(ParticleAction... actions) {
        this.actions = actions;
    }

    @Override
    public void tick() {
        super.tick();
        if (index < actions.length) {
            actions[index].tick();
        }
    }

    @Override
    public void operate(BaseActionParticle particle, float partialTick) {
        super.operate(particle, partialTick);

        // Shift to next action if needed
        if (actions[index].isFinished()) {
            index++;
            if (index >= actions.length) {
                finished = true;
                return;
            }
            actions[index].beginAction(particle);
        }

        if (index < actions.length) {
            actions[index].operate(particle, partialTick);
        }
    }

    @Override
    public ParticleAction copy() {
        ParticleAction[] newActions = new ParticleAction[actions.length];
        for (int i = 0; i < actions.length; i++) {
            newActions[i] = actions[i].copy();
        }
        return new SequenceAction(newActions);
    }

    @Override
    public void beginAction(BaseActionParticle particle) {
        index = 0;
        actions[index].beginAction(particle);
    }

    @Override
    public void operateAction(BaseActionParticle particle, double time) {
        // Action does nothing
    }
}
